### PR TITLE
chore(dev): improve dev mode

### PR DESCRIPTION
**Pull request by sidvishnoi**
Lets see how this works out :smile: 
image

- Starts serve so one can play with examples/basic.html
- Starts karma server so when can go to localhost:9876/debug.html and have fun there.
- Starts a file watcher, which on change: creates a debug build and runs karma tests

Closes https://github.com/w3c/respec/issues/2559
cc: @marcoscaceres @saschanaz 

TODO:
- ~add command line args~
- ~fix that welcome message banner (it's very hacky, spent unnecessarily long time on it :sweat_smile: )... maybe we can use [boxen](https://www.npmjs.com/package/boxen) if we want to be fancy here~
- ~Maybe not always build and run tests (sometimes we just want to prototype with basic.html). Pass a CLI flag to do so.~ Added interactive mode
- ~run particular tests~ (supports --grep) or only those whose source changed? (maybe some other PR)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "@types/pluralize": "0.0.29",
     "babel-eslint": "^10.0.3",
     "chai": "^4.2.0",
+    "chokidar": "^3.3.0",
```
Comment (marcoscaceres) [inline package.json]: Interested to hear why you chose this one over the other ones?

File: tools/dev.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        box.vert,
+        colors.white.bold(`${msg[0]}:`),
+        colors.white(msg[1]),
+        box.vert,
+      ].join("\t");
+    }),
+    ` ${box.bottomLeft}${box.hr.repeat(hrWidth)}${box.bottomRight}`,
+  ];
+  console.log(colors.bgBlack.green(lines.map(line => `\t${line}`).join("\n")));
+  console.log();
```
Comment (marcoscaceres) [inline tools/dev.js]: Nit: would `console.log(colors.bgBlack.green(lines.map(line => `\t${line}`).join("\n")), "");` do the same? Then we don't need this empty one?

File: tools/dev.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    console.error(colors.error(err.stack));
+  }
+}
+
+function onError(err) {
+  console.error(colors.error(err.stack));
+  process.exit(1);
+}
+
+function printWelcomeMessage() {
```
Comment (marcoscaceres) [inline tools/dev.js]: 😁, lol slightly overkill... but ok.

File: tools/dev.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  {
+    browsers: ["ChromeHeadless"],
+    autoWatch: false,
+    port: KARMA_PORT,
+    logLevel: karma.constants.LOG_WARN,
+  }
+);
+const karmaServer = new karma.Server(karmaConfig);
+
+main();
```
Comment (marcoscaceres) [inline tools/dev.js]: Main looks kinda redundant, tbh... maybe we should just ditch main() and just inline all this?

File: tools/dev.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+const serveConfig = require("../serve.json");
+const { Builder } = require("./builder");
+
+const KARMA_PORT = 9876;
+const SERVE_PORT = 5000;
+
+const karmaConfig = karma.config.parseConfig(
+  path.join(__dirname, "../karma.conf.js"),
+  {
+    browsers: ["ChromeHeadless"],
```
Comment (saschanaz) [inline tools/dev.js]: But...but...👀

File: tools/dev.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  watcher.on("all", onFileChange);
+  watcher.on("error", onError);
+
+  printWelcomeMessage();
+}
+
+async function onFileChange(_event, _file) {
+  try {
+    await Builder.build({ name: "w3c", debug: true });
+    karma.runner.run({ port: KARMA_PORT, autoWatch: true }, () => {});
```
Comment (saschanaz) [inline tools/dev.js]: Can we cancel the previous run when file changes happen but the tests are not finished yet?

Comment (sidvishnoi) [reply tools/dev.js]: Couldn't figure that one out. Will try again

File: tools/dev.js (hunk 5 above)
Comment (sidvishnoi) [reply tools/dev.js]: This will be cli arg. Using this during prototyping as it doesn't open a new window

File: package.json (hunk 1 above)
Comment (sidvishnoi) [reply package.json]: Nodemon: API wasn't nice
Watch: too raw
Chokidar: nice API, have used previously, claims to be a wrapper that fixes node's fs events
(Will write a detailed one when I am on laptop)

File: tools/dev.js (hunk 5 above)
Comment (saschanaz) [reply tools/dev.js]: Yes, but...but... why not Firefox! 😂

File: tools/dev.js (hunk 2 above)
Comment (sidvishnoi) [reply tools/dev.js]: that empty one is for new line though, so will add a `"\n"` in end instead

Comment (saschanaz) [reply tools/dev.js]: And Marcos' code will insert a new line too!

File: tools/dev.js (hunk 6 above)
Comment (sidvishnoi) [reply tools/dev.js]: So the karma API allows to stop the server, but doesn't allow stopping a run in between. One way could be to stop the server so the run fails and restart the server again.

Comment (saschanaz) [reply tools/dev.js]: That sounds reasonable enough.

Comment (sidvishnoi) [reply tools/dev.js]: Didn't work :smile: 
Want to give it a try? https://karma-runner.github.io/4.0/dev/public-api.html

File: tools/dev.js (hunk 2 above)
Comment (sidvishnoi) [reply tools/dev.js]: first: with `"\n"`, second: with `""`
image
(talking about an empty line after box)

Comment (saschanaz) [reply tools/dev.js]: Ooops you're right, somehow `console.log("abc\n")` and `console.log("abc")` is equivalent, the last explicit newline is ignored. 🤯

File: tools/dev.js (hunk 6 above)
Comment (sidvishnoi) [reply tools/dev.js]: karmaServer is in EventEmitter :tada: 
``` js
karmaServer.emit("run_complete", [], {});
```

**Comment by sidvishnoi** [comment]
Added command line args to:
- build specific profile (`--profile PROFILE`)
- specify karma browsers (`--browsers FirefoxHeadless ChromeHeadless`)
- run in interactive mode (disables file watching, press `t` to build and run tests) (`--interactive`)

The interactive mode might not work on Windows (TTY needed) @saschanaz can you play with it?

**Comment by saschanaz** [comment]
> The interactive mode might not work on Windows (TTY needed) @saschanaz can you play with it?

Works great 👍

**Comment by wetneb** [comment]
This looks amazing! Speaking of which, [the wiki currently recommends](https://github.com/w3c/respec/wiki/Developers-Guide) `npm run babel:watch` which does not seem to work:
```
npm ERR! missing script: babel:watch
```

**Comment by saschanaz** [comment]
> This looks amazing! Speaking of which, [the wiki currently recommends](https://github.com/w3c/respec/wiki/Developers-Guide) `npm run babel:watch` which does not seem to work:

Oops, you don't need babel anymore, as we removed babel dependency and migrated to native ES module 👍

**Comment by wetneb** [comment]
> Oops, you don't need babel anymore, as we removed babel dependency and migrated to native ES module

OK, I'm not sure what to replace it with (or delete?) so I'm not editing it myself. I think there are a couple of other references to running `babel:` things in the rest of the wiki / repo which might be worth cleaning up too.

File: tools/dev-server.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  stdin.on("data", async key => {
+    switch (key) {
+      case "\u0003":
+      case "q": {
+        return karma.stopper.stop(karmaConfig, code => {
+          setTimeout(() => process.exit(code), 0);
+        });
+      }
+      case "t":
+        return await buildAndTest();
```
Comment (saschanaz) [inline tools/dev-server.js]: Seems that typing `t` multiple times causes building multiple time in parallel 👀

File: tools/dev-server.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      case "h":
+        return printWelcomeMessage(args);
+      default:
+        process.stdout.write(key);
+    }
+  });
+}
+
+async function onFileChange(_event, _file) {
+  await buildAndTest();
```
Comment (saschanaz) [inline tools/dev-server.js]: Could the karma.stopper also be used here?

File: tools/dev-server.js (hunk 7 above)
Comment (sidvishnoi) [reply tools/dev-server.js]: Added an active flag to let only single build run at a time. Try again?

File: tools/dev-server.js (hunk 8 above)
Comment (sidvishnoi) [reply tools/dev-server.js]: Not sure how. karma.stopper stops the server altogether, and based on my attempts, it fails to restart :confused:

**Pull request merged**